server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/GetMessageListMethodTest.java

```diff
@@ -1332,6 +1334,74 @@ public void getMessageListShouldComputeTextBodyWhenNoTextBodyButHtmlBody() throw
+                ClassLoader.getSystemResourceAsStream("eml/twoAttachments.eml"), new Date(), false, new Flags(Flags.Flag.FLAGGED));
+
+        await();
+
+        given()
+            .header("Authorization", accessToken.serialize())
+            .body("[[\"getMessageList\", {\"filter\":{\"keywords\":[\"$Flagged\"]}}, \"#0\"]]")
+        .when()
+            .post("/jmap")
+        .then()
+            .statusCode(200)
+            .body(NAME, equalTo("messageList"))
+            .body(ARGUMENTS + ".messageIds", allOf(
+                containsInAnyOrder(messageFlagged.getMessageId().serialize()),
+                not(containsInAnyOrder(messageNotFlagged.getMessageId().serialize()))));
+    }
+
+    @Test
+    public void getMessageListKeywordsFilterShouldWorkWhenUserFlag() throws Exception {
+        mailboxProbe.createMailbox(MailboxConstants.USER_NAMESPACE, username, "mailbox");
+
+        Flags flags = FlagsBuilder.builder()
+            .add(Flags.Flag.FLAGGED)
+            .add(JmapFlag.Constants.FORWARDED)
+            .build();
+
+        ComposedMessageId messageNotFlagged = mailboxProbe.appendMessage(username, new MailboxPath(MailboxConstants.USER_NAMESPACE, username, "mailbox"),
+                new ByteArrayInputStream("Subject: test\r\n\r\ntestmail".getBytes()), new Date(), false, new Flags());
+        ComposedMessageId messageFlagged = mailboxProbe.appendMessage(username, new MailboxPath(MailboxConstants.USER_NAMESPACE, username, "mailbox"),
+                ClassLoader.getSystemResourceAsStream("eml/twoAttachments.eml"), new Date(), false, flags);
+
+        await();
+
+        given()
+            .header("Authorization", accessToken.serialize())
+            .body("[[\"getMessageList\", {\"filter\":{\"keywords\":[\"$Flagged\", \"$Forwarded\"]}}, \"#0\"]]")
+        .when()
+            .post("/jmap")
+        .then()
+            .statusCode(200)
+            .body(NAME, equalTo("messageList"))
+            .body(ARGUMENTS + ".messageIds", allOf(
+            containsInAnyOrder(messageFlagged.getMessageId().serialize()),
+            not(containsInAnyOrder(messageNotFlagged.getMessageId().serialize()))));
+    }
+
+    @Test
+    public void getMessageListKeywordsFilterShouldReturnAnErrorWhenKeywordsHaveUnknowJmapFlag() throws Exception {
+        given()
+            .header("Authorization", accessToken.serialize())
+            .body("[[\"getMessageList\", {\"filter\":{\"keywords\":[\"$Flagged\", \"UnknownValue\"]}}, \"#0\"]]")
+        .when()
+            .post("/jmap")
+        .then()
+            .statusCode(200)
+            .body(NAME, equalTo("messageList"))
+            .body(NOT_VALID_ARGUMENTS, hasSize(1))
+            .body(NOT_VALID_ARGUMENTS + "[0].type", equalTo("invalidArguments"))
+            .body(NOT_VALID_ARGUMENTS + "[0].properties[0]", equalTo("keywords"));
+    }
+
     private Date convertToDate(LocalDate localDate) {
```

Comment:
We can search by $Draft but ignore Recent and Deleted by filter it before searching